### PR TITLE
nixos/httpd: add virtualHosts.<name>.listenAddresses option

### PR DESCRIPTION
**Pull request by aanderse**
###### Motivation for this change
- https://github.com/NixOS/nixpkgs/pull/131962
- [Keeping up with the Joneses](https://en.wikipedia.org/wiki/Keeping_up_with_the_Joneses)

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/web-servers/apache-httpd/vhost-options.nix [hunk 1]
```diff
@@ -47,10 +47,27 @@ in
...
+
+    listenAddresses = mkOption {
+      type = with types; listOf str;
+
+      description = ''
+        Listen addresses for this virtual host.
+        Compared to <literal>listen</literal> this only sets the addreses
+        and the ports are choosen automatically.
       '';
+      default = [];
```
Comment (roberth) [inline nixos/modules/services/web-servers/apache-httpd/vhost-options.nix:69]:
```suggestion
      default = [ "*" ];
```

File: nixos/modules/services/web-servers/apache-httpd/default.nix [hunk 2]
```diff
@@ -36,10 +36,13 @@ let
...
+      addrs = if hostOpts.listenAddresses != [] then hostOpts.listenAddresses else [ "*" ];
+    in
     if hostOpts.listen != [] then hostOpts.listen
-    else (
-      optional (hostOpts.onlySSL || hostOpts.addSSL || hostOpts.forceSSL) { ip = "*"; port = 443; ssl = true; } ++
-      optional (!hostOpts.onlySSL) { ip = "*"; port = 80; ssl = false; }
+    else flatten (
+      optional (hostOpts.onlySSL || hostOpts.addSSL || hostOpts.forceSSL) (map (addr: { ip = addr; port = 443; ssl = true; }) addrs) ++
+      optional (!hostOpts.onlySSL) (map (addr: { ip = addr; port = 80; ssl = false; }) addrs)
     );
```
Comment (roberth) [inline nixos/modules/services/web-servers/apache-httpd/default.nix:46]:
```suggestion
    else
      optionals (hostOpts.onlySSL || hostOpts.addSSL || hostOpts.forceSSL) (map (addr: { ip = addr; port = 443; ssl = true; }) addrs) ++
      optionals (!hostOpts.onlySSL) (map (addr: { ip = addr; port = 80; ssl = false; }) addrs);
```

This should be equivalent.

File: nixos/modules/services/web-servers/apache-httpd/vhost-options.nix [hunk 3]
```diff
@@ -47,10 +47,27 @@ in
...
+        </para>
+        <para>
+          If you only want to set the addresses manually and not the ports, take a look at <literal>listenAddresses</literal>.
+        </para>
+        </note>
+      '';
+    };
+
+    listenAddresses = mkOption {
+      type = with types; listOf str;
```
Comment (roberth) [inline nixos/modules/services/web-servers/apache-httpd/vhost-options.nix:62]:
```suggestion
      type = with types; nonEmptyListOf str;
```

In conjunction with the changed default below.

File: nixos/modules/services/web-servers/apache-httpd/default.nix [hunk 4]
```diff
@@ -36,10 +36,13 @@ let
   dependentCertNames = unique (map (hostOpts: hostOpts.certName) acmeEnabledVhosts);
 
   mkListenInfo = hostOpts:
+    let
+      addrs = if hostOpts.listenAddresses != [] then hostOpts.listenAddresses else [ "*" ];
```
Comment (roberth) [inline nixos/modules/services/web-servers/apache-httpd/default.nix:40]:
```suggestion
      addrs = hostOpts.listenAddresses;
```

In conjunction with option change below.
(and you can inline `addrs` of course)

**Review by roberth — COMMENTED** [review]
Just some nixlang suggestions. I can't say much about httpd, but it doesn't seem complicated.

**Comment by roberth** [comment]
Heh, github "Conversation" items and "Files changed" items appear in different order. I'll make better use of that in the future. My comments were written for the "Files changed" view.

File: nixos/modules/services/web-servers/apache-httpd/default.nix (hunk 2 above)
Comment (aanderse) [reply nixos/modules/services/web-servers/apache-httpd/default.nix:46]: I tested this out and without `flatten` a list of lists was produced... :thinking:

File: nixos/modules/services/web-servers/apache-httpd/vhost-options.nix (hunk 3 above)
Comment (aanderse) [reply nixos/modules/services/web-servers/apache-httpd/vhost-options.nix:62]: Awesome! I think I recall seeing `nonEmptyListOf` being introduced, but completely forgot about it. Much appreciated.

File: nixos/modules/services/web-servers/apache-httpd/default.nix (hunk 2 above)
Comment (roberth) [reply nixos/modules/services/web-servers/apache-httpd/default.nix:46]: Even with plural `optional` -> `optionals`?

File: nixos/modules/services/web-servers/apache-httpd/vhost-options.nix (hunk 1 above)
Comment (aanderse) [reply nixos/modules/services/web-servers/apache-httpd/vhost-options.nix:69]: I actually had a branch that did some cleanup like this, but never ended up pushing it... With this PR I was going for as much copy/paste of #131962 as possible, but you're right, better is better :smile: I'll take a look at that.

File: nixos/modules/services/web-servers/apache-httpd/default.nix (hunk 2 above)
Comment (aanderse) [reply nixos/modules/services/web-servers/apache-httpd/default.nix:46]: Oh... github should highlight that extra `s` :laughing: Let me look at that again.

**Comment by aanderse** [comment]
@roberth I'm always excited to have you leave a review on my PRs. I'll take some time to apply your changes and report back. Thanks!

**Comment by aanderse** [comment]
@roberth should be good now.

File: nixos/modules/services/web-servers/apache-httpd/vhost-options.nix [hunk 5]
```diff
@@ -47,10 +47,27 @@ in
...
+      '';
+    };
+
+    listenAddresses = mkOption {
+      type = with types; nonEmptyListOf str;
+
+      description = ''
+        Listen addresses for this virtual host.
+        Compared to <literal>listen</literal> this only sets the addreses
+        and the ports are choosen automatically.
```
Comment (roberth) [inline nixos/modules/services/web-servers/apache-httpd/vhost-options.nix:67]:
```suggestion
        and the ports are chosen automatically.
```

**Review by roberth — APPROVED** [review]
lgtm. I suppose you've tested this manually?

You could test a virtual host with loopback only in the nixosTest to confirm.

**Comment by aanderse** [comment]
> lgtm. I suppose you've tested this manually?

I have. I quickly tested this on a few different configurations I had and found the generated configuration to be equivalent. Thanks for your review and suggestions :heart:

**Comment by roberth** [comment]
I see we don't have a test for apache httpd specifically.

@GrahamcOfBorg test acme

**Pull request merged**